---
title: "SVCFit Guide"
subtitle: 
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SVCFit Guide}
  %\VignetteEngine{knitr::knitr}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, echo=TRUE, message=FALSE}
library(SVCFit)
library(tidyverse)
library(ggplot2)
library(gridExtra)
library(igraph)
```

# Introduction

This guide provides a walkthrough of the SVCFit package, demonstrating its functionality and usage. We begin by showcasing an example of the input data format and how different functions within SVCFit process and modify it. Next, we present two additional sample datasets to illustrate how SVCFit's output can be visualized and interpreted. 


# Input to SVCFit

This is an example VCF file that would be generated by Manta [1] under *tumor_only* mode, so it doesn't have a column for reads from matched normal sample.

```{r,echo = FALSE}
file_path <- system.file("extdata", "example_sv.vcf", package = "SVCFit")
example_vcf <- read.table(file_path, quote="\"")
colnames(example_vcf)=c("CHROM", "POS", "ID", "REF", "ALT", "QUAL", "FILTER", "INFO", "FORMAT", "normal", "tumor")
```

```{r, echo=TRUE}
head(as_tibble(example_vcf))
```

This input data has `r nrow(example_vcf)` structural variants. Given this input data, let's see how SVCFit works!

# General workflow

## 1. Extract information from original VCF (extract_info)

This step assigns column names and extracts essential information for subsequent calculations and filtering. The key details extracted include reads, structural variant length, and coordinates. The command below includes `tumor_only=TRUE` since the input is generated without a matched normal sample, and sets `length_threshold=0` to include structural variants with lengths greater than 0.

```{r, message=FALSE}
p_het=system.file("extdata", "example/het_near_sv_c50p80m50.vcf", package = "SVCFit")
p_onsv=system.file("extdata", "example/het_on_sv_c50p80m50.vcf", package = "SVCFit")
p_sv=system.file("extdata", "example/example_sv.bed", package = "SVCFit")
p_cnv=system.file("extdata", "example/c50p80m50.bed", package = "SVCFit")
t_only=FALSE

info=extract_info(p_het, p_onsv, p_sv, p_cnv, chr=NULL, flank_del=50, QUAL_tresh=100, min_alt=2, t_only)

  data <- info[[1]]
  sv_info <- info[[2]]
  snp_df <- info[[3]]
  sv_phase <- info[[4]]
```

## 2. Infer Zygosity and phasing

```{r}
anno_sv_cnv <- characterize_sv(sv_phase, sv_info, data[[4]])

```


## 3. Calculate SVCF for structural variants

This step calculates the structural variant cellular fraction (SVCF) for all structural variants in the given VCF file.

```{r, echo = TRUE}
svcfit_out <- calc_svcf(anno_sv_cnv, sv_info, thresh=0.1, samp, exper)
```


## Additional functions

`load_truth` and `attach_truth` are only applicable to data from multiple-clone simulations where the clone assignment for each structural variants is known. In the example below, we see 4 structural variants included in this simulation:

```{r, echo=FALSE}
overlap=ifelse(exper=='exp1', FALSE, TRUE)
  truth <- load_truth(p_truth, overlap)
  final_truth <- attach_truth(final, truth)
```