---
title: "Figure_1"
author: "Yunzhou Liu"
date: "2025-11-21"
output: html_document
---

```{r setup, include=FALSE, message=FALSE}
#library(SVCFit)
library(dplyr)
library(tidyr)
library(purrr)
library(stringr)
library(GenomicRanges)
library(ggplot2)
library(ggpubr) #for stat
library(rstatix) #for stat
library(ggsignif) #for plot significance
mendeley_dir = '~/Karchin Lab Dropbox/YunZhou Liu/SVCFit-2024/NPJ_Precision_Oncology_submission/mendeley_data'
```

# Fig 1 VISOR benchmark
```{r, message=FALSE}
## can directly load data from /int/extdata/VISOR_bench_data.bed
analyze <- function(samp, exper, p_het, p_onsv, p_sv, p_cnv, p_truth, thresh=0.1, chr=NULL, flank_del=50, QUAL_tresh=100, min_alt=2, t_only=FALSE) {
  info <- extract_info(p_het, p_onsv, p_sv, p_cnv, chr=NULL, flank_del=50, QUAL_tresh=100, min_alt=2, t_only)
  data <- info[[1]]
  sv_info <- info[[2]]
  snp_df <- info[[3]]
  sv_phase <- info[[4]]
  anno_sv_cnv <- characterize_sv(sv_phase, sv_info, data[[4]])
  final <- calc_svcf(anno_sv_cnv, sv_info, thresh=0.1, samp, exper)
  overlap=ifelse(exper=='exp1', FALSE, TRUE)
  truth <- load_truth(p_truth, overlap)
  final_truth <- attach_truth(final, truth)
  svc <- analyze_svclone(samp,exper, final_truth, sv_info)
  common_dat <- final_truth %>%
    filter(ID %in% svc$ID)
  return(list(final_truth, svc, common_dat))
}

condition_list=c(
  "c50p20m10", "c50p20m30", "c50p20m50",
  "c50p40m10", "c50p40m30", "c50p40m50",
  "c50p60m10", "c50p60m30", "c50p60m50",
  "c50p80m10", "c50p80m30", "c50p80m50"
)
base_dir = paste0(mendeley_dir, '/VISOR_benchmark')

e1 = lapply(condition_list, 
             function(x) analyze(x,'exp1',
                                 file.path(base_dir, 'SNP', 'exp1', paste0("het_near_sv_",x, ".vcf")),
                                 file.path(base_dir, 'SNP', 'exp1', paste0("het_on_sv_",x,".vcf")),
                                 file.path(base_dir, 'SV_vcfs', 'exp1', paste0('svt_', x, '.vcf')),
                                 file.path(base_dir, 'facet', 'exp1', paste0(x, '.bed')),
                                 file.path(base_dir, 'truth')))
e2 = lapply(condition_list, 
             function(x) analyze(x,'exp2',
                                 file.path(base_dir, 'SNP', 'exp2', paste0("het_near_sv_",x, ".vcf")),
                                 file.path(base_dir, 'SNP', 'exp2', paste0("het_on_sv_",x,".vcf")),
                                 file.path(base_dir, 'SV_vcfs', 'exp2', paste0('svt_', x, '.vcf')),
                                 file.path(base_dir, 'facet', 'exp2', paste0(x, '.bed')),
                                 file.path(base_dir, 'truth')))
e3 = lapply(condition_list, 
             function(x) analyze(x,'exp3',
                                 file.path(base_dir, 'SNP', 'exp3', paste0("het_near_sv_",x, ".vcf")),
                                 file.path(base_dir, 'SNP', 'exp3', paste0("het_on_sv_",x,".vcf")),
                                 file.path(base_dir, 'SV_vcfs', 'exp3', paste0('svt_', x, '.vcf')),
                                 file.path(base_dir, 'facet', 'exp3', paste0(x, '.bed')),
                                 file.path(base_dir, 'truth')))
e4 = lapply(condition_list, 
             function(x) analyze(x,'exp4',
                                 file.path(base_dir, 'SNP', 'exp4', paste0("het_near_sv_",x, ".vcf")),
                                 file.path(base_dir, 'SNP', 'exp4', paste0("het_on_sv_",x,".vcf")),
                                 file.path(base_dir, 'SV_vcfs', 'exp4', paste0('svt_', x, '.vcf')),
                                 file.path(base_dir, 'facet', 'exp4', paste0(x, '.bed')),
                                 file.path(base_dir, 'truth')))
e5 = lapply(condition_list, 
             function(x) analyze(x,'exp5',
                                 file.path(base_dir, 'SNP', 'exp5', paste0("het_near_sv_",x, ".vcf")),
                                 file.path(base_dir, 'SNP', 'exp5', paste0("het_on_sv_",x,".vcf")),
                                 file.path(base_dir, 'SV_vcfs', 'exp5', paste0('svt_', x, '.vcf')),
                                 file.path(base_dir, 'facet', 'exp5', paste0(x, '.bed')),
                                 file.path(base_dir, 'truth')))

gather_data=function(lst, idx){
  test=lst[[idx]]
  svclone=as.data.frame(map_dfr(test, 2))
  com_svcf=as.data.frame(map_dfr(test, 3))
  tsvf=com_svcf %>% mutate(tool="svcfit", svcf=final_svcf) %>% select(tool,svcf, expmt, zygosity, classification, sample, clone)
  tsvc=svclone %>% mutate(tool="svclone")%>% select(tool,svcf, expmt, zygosity, classification, sample,clone)
  output=rbind(tsvf, tsvc)
  return(output)
}

ll=list(e1,e2,e3,e4,e5)
all_dat=lapply(1:5, function(x) gather_data(ll,x)) %>% 
  do.call(rbind,.)%>%
  mutate(pur=as.integer(gsub(".*p(\\d+).*","\\1",sample))/100,
         purity=paste0('purity ', pur*100,"%"),
         mixture=as.integer(gsub(".*m","",sample))/100,
         true_svcf=pur,
         true_svcf=case_when(
           clone=='clonal' ~ pur,
           clone=='2' ~ pur*mixture,
           clone=='sub' ~ pur*(1-mixture)
         ),
         new_class=paste0(zygosity, "_", classification),
         clone=case_when(
           clone=='2' ~ 'subclone1',
           clone=='sub' ~ 'subclone2',
           clone=='clonal' ~ 'clonal'
         ),
         error=abs(svcf-true_svcf))

all_dat$tool <- factor(all_dat$tool, levels = c("svcfit", "svclone"))

plot_dat <- all_dat %>%
  mutate(clonality  = if_else(clone == "clonal", "clonal", "subclonal"),
         purity = factor(purity, levels = c("purity 80%", "purity 40%"))) %>%
  filter(purity %in% c("purity 80%", "purity 40%")) %>%
  group_by(classification, clonality, tool, purity) %>%
  summarise(mean_err = mean(abs(true_svcf - svcf), na.rm = TRUE),
            .groups  = "drop") %>%
  mutate(sv = paste0(clonality, " ", classification),
         sv = factor(sv, levels = c(
        "subclonal INV",    "clonal INV",
        "subclonal DUP",    "clonal DUP",
        "subclonal DEL",    "clonal DEL",
        "subclonal BND",    "clonal BND")),
        classification = case_when(
          classification == 'INV' ~ 'inversion',
          classification == 'DUP' ~ 'tandem duplication',
          classification == 'DEL' ~ 'deletion',
          classification == 'BND' ~ 'translocation'
        ),
        classification = factor(classification,
                            levels = c("translocation","tandem duplication","inversion","deletion")),
        y = as.numeric(classification) + if_else(clonality == "clonal",  0.2, -0.2))

f1b_stat=all_dat %>%
  mutate(clonality  = if_else(clone == "clonal", "clonal", "subclonal"),
         purity = factor(purity, levels = c("purity 80%", "purity 40%")),
         zygosity=ifelse(zygosity=='het', 'heterozygous', 'homozygous'),
         error=(abs(svcf-true_svcf))) %>%
  filter(purity %in% c("purity 80%", "purity 40%")) %>%
  group_by(zygosity)%>%
  wilcox_test(error~tool, paired = TRUE)%>%
  add_significance()%>%
  add_xy_position(x = "zygosity")
```

## plot F1A,B
```{r pressure, echo=FALSE, message=FALSE}
f1a=ggplot(plot_dat, aes(x = mean_err, y = y)) +
  geom_line(aes(group = sv, linetype = clonality)) +
  geom_point(aes(color = tool), size = 3) +
  facet_grid(purity ~ .) +
  scale_y_continuous(
    breaks = seq_len(nlevels(plot_dat$classification)),
    labels = levels(plot_dat$classification),
    limits = c(0.5, nlevels(plot_dat$classification) + 0.5)
  ) +
  theme_bw()+
  theme(text = element_text(size = 15),
        strip.background = element_rect(fill = "white", color = "black", size = 0.3))+
  ylab('')+
  xlab('Mean absolute SVCF estimation error')


f1b=all_dat %>%
  mutate(clonality  = if_else(clone == "clonal", "clonal", "subclonal"),
         purity = factor(purity, levels = c("purity 80%", "purity 40%")),
         zygosity=ifelse(zygosity=='het', 'heterozygous', 'homozygous')) %>%
  filter(purity %in% c("purity 80%", "purity 40%")) %>%
  ggplot()+
  geom_violin(aes(x=zygosity, y=(abs(svcf-true_svcf)), fill = tool))+
  stat_pvalue_manual(f1b_stat, hide.ns = TRUE)+
  ylim(c(0,1.2))+
  ylab('Absolute SVCF estimation error')+
  xlab('')+
  theme_minimal()+
  theme(text = element_text(size = 15))

```


## define functions for F1C
```{r}
load_data <- function(samp, exper, base_dir) {
  dir <- file.path(base_dir)

  het_snp <- read.table(file.path(dir, paste0("SNP/het_near_sv_",exper, ".vcf")), quote="\"")
  colnames(het_snp) = c('CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','bulk')
  het_snp=het_snp %>% 
    filter(CHROM %in% c(1:22)) %>% 
    mutate(CHROM = ifelse(grepl('chr', CHROM), CHROM, paste0('chr',CHROM)))
  
  het_on_sv <- read.table(file.path(dir, paste0("SNP/het_on_sv_",exper,".vcf")), quote = "\"")
  colnames(het_on_sv) = c('CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','bm','gm')
  het_on_sv=het_on_sv %>% 
    filter(CHROM %in% c(1:22)) %>% 
    mutate(CHROM = ifelse(grepl('chr', CHROM), CHROM, paste0('chr',CHROM)))
  
  sv <- read.table(file.path(dir, paste0('prostate_vcfs/svt_', exper,'.vcf')), quote = '"')
  colnames(sv) = c('CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','normal','tumor')
  sv=sv %>% 
    filter(CHROM %in% c(1:22)) %>% 
    mutate(CHROM = ifelse(grepl('chr', CHROM), CHROM, paste0('chr',CHROM)))
  
  cnv <- read.delim(file.path(dir, paste0("facet/", exper, '.bed')), header = TRUE) %>%
    mutate(lcn.em=ifelse(is.na(lcn.em), 1, lcn.em),
           chrom=paste0("chr",chrom))%>%
    filter(chrom %in% paste0('chr',c(1:22)))

  return(list(het_snp = het_snp,
              het_on_sv = het_on_sv,
              sv = sv,
              cnv = cnv))
}

extract_info <- function(samp, exper, dir, chr_lst=NULL, flank_del=50, QUAL_tresh=100, min_alt=2, tumor_only=FALSE){
  data      <- load_data(samp, exper, dir)
  bnd_info   <- proc_bnd(data$sv, flank_del=50)
  sv_info   <- parse_sv_info(data$sv, bnd_info[[1]], bnd_info[[2]], QUAL_tresh=100, min_alt=2)
  snp_df    <- parse_het_snps(data$het_snp)
  sv_phase  <- parse_snp_on_sv(data$het_on_sv, snp_df)
  return(list(data, sv_info, snp_df, sv_phase))
}

parse_snp_on_sv <- function(het_on_sv, snp_df) {

  ########## new part ############
  sum_like_strings <- function(a, b) {
    # extract numbers
    nums_a <- as.numeric(str_extract_all(a, "\\d+")[[1]])
    nums_b <- as.numeric(str_extract_all(b, "\\d+")[[1]])

    # sum numbers
    summed <- nums_a + nums_b

    # extract non-numeric delimiters
    delims <- str_extract_all(a, "[^0-9]+")[[1]]

    # rebuild the string by interleaving
    out <- paste0(
      paste(mapply(paste0, summed, c(delims, "")), collapse = "")
    )
    out
  }

  het_on_sv$bulk <- mapply(sum_like_strings, het_on_sv$bm, het_on_sv$gm)
  #########################

  sv_phase <- het_on_sv %>%
    select(-ALT)%>%
    # get allele count for SNPs
    mutate(snp_id=row_number(),
           AD=gsub(".*:","",bulk),
           len=nchar(gsub("\\d","",AD)),
           DEP=gsub("DP=(\\d+);.*","\\1", INFO),
           onsv_ref=as.integer(gsub(",.*","", AD)),
           onsv_alt=as.integer(ifelse(len==2, gsub(".*,(\\d+),.*","\\1", AD),0)),
           onsv_alt2=as.integer(ifelse(len==2, gsub(".*,","\\1", AD), 0)), # check if this SNP has three alleles, which is filtered out
           a_count=(onsv_ref!=0) + (onsv_alt!=0),
           allele=ifelse(a_count==1 & onsv_ref!=0, REF, "other"))%>% # other means the allele is not reference allele (used to tell mat or pat origin)
    filter(DEP > 0,
           onsv_alt2==0)%>%
    # join in all het SNP
    left_join(snp_df, by = c("CHROM", "POS", "REF"))%>%
    select(-ID)

  return(sv_phase)
}


analyze <- function(samp, exper, dir, flank_del=50, QUAL_tresh=100, min_alt=2, t_only=FALSE) {
  info <- extract_info(samp, exper, dir, flank_del=50, QUAL_tresh=100, min_alt=2, t_only)
  data <- info[[1]]
  sv_info <- info[[2]]
  snp_df <- info[[3]]
  sv_phase <- info[[4]]
  anno_sv_cnv <- characterize_sv(sv_phase, sv_info, data$cnv)
  final <- calc_svcf(anno_sv_cnv, sv_info, thresh=0.1, samp, exper)
  return(final)
}
```


# process for F1 prostate mixture
```{r}
### load the truth
base_dir = paste0(mendeley_dir, '/Prostate_mixture')
tbm = read.table(paste0(base_dir, "/in_silico_true/bM.vcf"), quote="\"")
tgm = read.table(paste0(base_dir, "/in_silico_true/gM.vcf"), quote="\"")
colnames(tbm) = c('CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','normal','tumor')
colnames(tgm) = c('CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','normal','tumor')
bm = tbm %>%
  filter(FILTER == "PASS",
         !grepl('X', ALT),
         !grepl('Y', ALT),
         !grepl('hs37d5', ALT),
         !grepl('GL', ALT),
         CHROM %in% c(1:22))%>%
  mutate(samp='bm',
         chr2=ifelse(grepl('BND', ID), 
                     gsub('\\D+(\\d+):(\\d+).*','\\1', ALT), 
                     CHROM),
         pos2=ifelse(grepl('BND', ID), 
                     gsub('\\D+(\\d+):(\\d+).*','\\2', ALT), 
                     gsub('.*END=(\\d+);.*','\\1',INFO)),
         pos2=as.integer(pos2),
         SVID=ifelse(grepl('INV', ID), gsub("((:[^:\\D:]){5}$)","", ID), gsub("((:[^:\\D:]){4}$)","", ID)))%>%
  distinct(SVID, .keep_all = T)

gm = tgm %>%
  filter(FILTER == "PASS",
         !grepl('X', ALT),
         !grepl('Y', ALT),
         !grepl('hs37d5', ALT),
         !grepl('GL', ALT),
         CHROM %in% c(1:22))%>%
  mutate(samp='gm',
         chr2=ifelse(grepl('BND', ID), 
                     gsub('\\D+(\\d+):(\\d+).*','\\1', ALT), 
                     CHROM),
         pos2=ifelse(grepl('BND', ID), 
                     gsub('\\D+(\\d+):(\\d+).*','\\2', ALT), 
                     gsub('.*END=(\\d+);.*','\\1',INFO)),
         pos2=as.integer(pos2),
         SVID=ifelse(grepl('INV', ID), gsub("((:[^:\\D:]){5}$)","", ID), gsub("((:[^:\\D:]){4}$)","", ID)))%>%
  distinct(SVID, .keep_all = T)

clonal = bm %>%
  rowwise()%>%
  mutate(type=any(CHROM==gm$CHROM & chr2==gm$chr2 & abs(POS-gm$POS)<50 & abs(pos2-gm$pos2)<50))%>%
  filter(type)%>%
  mutate(clone='clonal')
subb = bm %>%
  rowwise()%>%
  mutate(type=any(CHROM==gm$CHROM & chr2==gm$chr2 & abs(POS-gm$POS)<50 & abs(pos2-gm$pos2)<50))%>%
  filter(!type)%>%
  mutate(clone='bm')
subg = gm %>%
  rowwise()%>%
  mutate(type=any(CHROM==bm$CHROM & chr2==bm$chr2 & abs(POS-bm$POS)<50 & abs(pos2-bm$pos2)<50))%>%
  filter(!type)%>%
  mutate(clone='gm')

all=rbind(clonal, subb, subg) %>% ungroup() %>% mutate(CHROM = paste0('chr',CHROM))

rm(bm, gm, subb, subg, clonal, tbm, tgm)

############## now load svcfit data
exp_list=c("3m19","3m28","3m37","3m46","3m55","3m64","3m73","3m82","3m91")
prost = lapply(exp_list, function(x) analyze('prostate', x,base_dir, 0.1))

svcfit=prost %>%
  do.call(rbind,.)%>%
  mutate(bmp=as.integer(gsub("3m(\\d)(\\d)","\\1", expmt))/10,
         gmp=as.integer(gsub("3m(\\d)(\\d)","\\2", expmt))/10)%>%
  rowwise()%>%
  mutate(type=any(CHROM==all$CHROM & abs(POS-all$POS)<50))%>%
  filter(type)%>%
  mutate(row=which(CHROM==all$CHROM & abs(POS-all$POS)<50)[1],
         clone=all$clone[row],
         ccf=final_svcf/0.475,
         ccf=ifelse(ccf>1, 1, ccf))%>%
  ungroup()%>%
  mutate(truth=case_when(
    clone=='clonal' ~ 1,
    clone=='bm' ~ bmp,
    clone=='gm' ~ gmp),
    error=truth-ccf)%>%
  #filter(!is.na(tascn))%>%
  group_by(expmt)%>%
  summarize(mean_error=abs(mean(error)))%>%
  mutate(method="svcfit")


# 4 and 5 clusters---------
exp_list=c('4m','5m')
prost45 = lapply(exp_list, function(x) analyze('prostate', x, base_dir, 0.1))

svcfit45=prost45 %>%
  do.call(rbind,.)%>%
  mutate(chr=as.integer(gsub('chr','',CHROM)),
         method="svcfit")%>%
  rowwise()%>%
  mutate(type=any(CHROM==all$CHROM & abs(POS-all$POS)<50))%>%
  filter(type)%>%
  mutate(row=which(CHROM==all$CHROM & abs(POS-all$POS)<50)[1],
         clone=all$clone[row])%>%
  ungroup() %>%
  mutate(truth=case_when(
    expmt=='4m' & chr%%2==0 & clone=='bm' ~ 0.6,
    expmt=='4m' & chr%%2==1 & clone=='bm' ~ 0.2,
    expmt=='4m' & chr%%2==1 & clone=='gm' ~ 0.8,
    expmt=='4m' & clone=='clonal' ~ 1,
    expmt=='5m' & chr%%2==0 & clone=='bm' ~ 0.6,
    expmt=='5m' & chr%%2==1 & clone=='bm' ~ 0.8,
    expmt=='5m' & chr%%2==1 & clone=='gm' ~ 0.2,
    expmt=='5m' & chr%%2==0 & clone=='gm' ~ 0.4,
    expmt=='5m' & clone=='clonal' ~ 1,
  ),
  ccf=final_svcf/0.475,
  ccf=ifelse(ccf>1, 1, ccf),
  error=truth-ccf)%>%
  filter(!is.na(truth))%>%
  #filter(!is.na(tascn))%>%
  group_by(expmt)%>%
  summarize(mean_error=abs(mean(error)))%>%
  mutate(method="svcfit")

svclone <- c(0.095, 0.095, 0.090, 0.103, 0.105, 0.107, 0.093, 0.109, 0.105, 0.075, 0.090)
name <- c('3m19','3m28','3m37','3m46','3m55','3m64','3m73','3m82','3m91','4m','5m')
paper_dat=data.frame(expmt=name, mean_error=svclone) %>%
  mutate(method="svclone")

x_lab <- c('10-90','20-80','30-70','40-60','50-50','60-40','70-30','80-20','90-10','4 clust','5 clust')

final=rbind(svcfit, svcfit45, paper_dat)

f1c=final %>%
  ggplot()+
  geom_point(aes(x=expmt, y = mean_error, color=method))+
  geom_line(aes(x=expmt, y = mean_error, group = method, color= method))+
  ylim(c(-0.5, 0.5))+
  ylab("Mean absolute CCF estimation error")+
  xlab("")+
  scale_x_discrete(labels = x_lab)+
  theme_minimal()+
  theme(text = element_text(size = 15),
        axis.text.x = element_text(angle = 45, hjust = 1))

```