---
title: "Figure_1"
author: "Yunzhou Liu"
date: "2025-11-21"
output: html_document
---

```{r setup, include=FALSE}
library(SVCFit)
library(tidyverse)
library(ggpubr) #for stat
library(rstatix) #for stat
library(ggsignif) #for plot significance
```

# F1 VISOR benchmark
```{r}
analyze <- function(p_het, p_onsv, p_sv, p_cnv, flank_del=50, QUAL_tresh=100, min_alt=2, tumor_only=FALSE, samp, exper) {
  info <- extract_info(p_het, p_onsv, p_sv, p_cnv, flank_del=50, QUAL_tresh=100, min_alt=2, tumor_only=FALSE)
  data <- info[[1]]
  sv_info <- info[[2]]
  snp_df <- info[[3]]
  sv_phase <- info[[4]]
  anno_sv_cnv <- characterize_sv(sv_phase, sv_info, data$cnv)
  final <- calc_svcf(anno_sv_cnv, sv_info, thresh=0.1, samp, exper)
  truth <- load_truth(exper)
  final_truth <- attach_truth(final, truth)
  svc <- analyze_svclone(samp,exper, final_truth, sv_info)
  common_dat <- final_truth %>%
    filter(ID %in% svc$ID)
  return(list(final_truth, svc, common_dat))
}

exp_list=c("exp1","exp2","exp3", "exp4","exp5")

p81 = lapply(exp_list, function(x) analyze("c50p80m10",x,
                                           file.path(base_dir, 'SNP', "c50p80m10", x, paste0("het_near_sv_",x, ".vcf")),
                                           file.path(base_dir, 'SNP', "c50p80m10", x, paste0("het_on_sv_",x,".vcf")),
                                           file.path(base_dir, 'svtyp', "c50p80m10", x, paste0('svt_', x, '.vcf')),
                                           file.path(base_dir, 'facet', "c50p80m10", x, paste0(x, '.bed')),
                                           0.1, chr_list = NULL))
p83 = lapply(exp_list, function(x) analyze("c50p80m30",x,
                                           file.path(base_dir, 'SNP', "c50p80m30", x, paste0("het_near_sv_",x, ".vcf")),
                                           file.path(base_dir, 'SNP', "c50p80m30", x, paste0("het_on_sv_",x,".vcf")),
                                           file.path(base_dir, 'svtyp', "c50p80m30", x, paste0('svt_', x, '.vcf')),
                                           file.path(base_dir, 'facet', "c50p80m30", x, paste0(x, '.bed')),
                                           0.1, chr_list = NULL))
p85 = lapply(exp_list, function(x) analyze("c50p80m50",x,
                                           file.path(base_dir, 'SNP', "c50p80m50", x, paste0("het_near_sv_",x, ".vcf")),
                                           file.path(base_dir, 'SNP', "c50p80m50", x, paste0("het_on_sv_",x,".vcf")),
                                           file.path(base_dir, 'svtyp', "c50p80m50", x, paste0('svt_', x, '.vcf')),
                                           file.path(base_dir, 'facet', "c50p80m50", x, paste0(x, '.bed')),
                                           0.1, chr_list = NULL))
p61 = lapply(exp_list, function(x) analyze("c50p60m10",x,
                                           file.path(base_dir, 'SNP', "c50p60m10", x, paste0("het_near_sv_",x, ".vcf")),
                                           file.path(base_dir, 'SNP', "c50p60m10", x, paste0("het_on_sv_",x,".vcf")),
                                           file.path(base_dir, 'svtyp', "c50p60m10", x, paste0('svt_', x, '.vcf')),
                                           file.path(base_dir, 'facet', "c50p60m10", x, paste0(x, '.bed')),
                                           0.1, chr_list = NULL))
p63 = lapply(exp_list, function(x) analyze("c50p60m30",x,
                                           file.path(base_dir, 'SNP', "c50p60m30", x, paste0("het_near_sv_",x, ".vcf")),
                                           file.path(base_dir, 'SNP', "c50p60m30", x, paste0("het_on_sv_",x,".vcf")),
                                           file.path(base_dir, 'svtyp', "c50p60m30", x, paste0('svt_', x, '.vcf')),
                                           file.path(base_dir, 'facet', "c50p60m30", x, paste0(x, '.bed')),
                                           0.1, chr_list = NULL))
p65 = lapply(exp_list, function(x) analyze("c50p60m50",x,
                                           file.path(base_dir, 'SNP', "c50p60m50", x, paste0("het_near_sv_",x, ".vcf")),
                                           file.path(base_dir, 'SNP', "c50p60m50", x, paste0("het_on_sv_",x,".vcf")),
                                           file.path(base_dir, 'svtyp', "c50p60m50", x, paste0('svt_', x, '.vcf')),
                                           file.path(base_dir, 'facet', "c50p60m50", x, paste0(x, '.bed')),
                                           0.1, chr_list = NULL))
p41 = lapply(exp_list, function(x) analyze("c50p40m10",x,
                                           file.path(base_dir, 'SNP', "c50p40m10", x, paste0("het_near_sv_",x, ".vcf")),
                                           file.path(base_dir, 'SNP', "c50p40m10", x, paste0("het_on_sv_",x,".vcf")),
                                           file.path(base_dir, 'svtyp', "c50p40m10", x, paste0('svt_', x, '.vcf')),
                                           file.path(base_dir, 'facet', "c50p40m10", x, paste0(x, '.bed')),
                                           0.1, chr_list = NULL))
p43 = lapply(exp_list, function(x) analyze("c50p40m30",x,
                                           file.path(base_dir, 'SNP', "c50p40m30", x, paste0("het_near_sv_",x, ".vcf")),
                                           file.path(base_dir, 'SNP', "c50p40m30", x, paste0("het_on_sv_",x,".vcf")),
                                           file.path(base_dir, 'svtyp', "c50p40m30", x, paste0('svt_', x, '.vcf')),
                                           file.path(base_dir, 'facet', "c50p40m30", x, paste0(x, '.bed')),
                                           0.1, chr_list = NULL))
p45 = lapply(exp_list, function(x) analyze("c50p40m50",x,
                                           file.path(base_dir, 'SNP', "c50p40m50", x, paste0("het_near_sv_",x, ".vcf")),
                                           file.path(base_dir, 'SNP', "c50p40m50", x, paste0("het_on_sv_",x,".vcf")),
                                           file.path(base_dir, 'svtyp', "c50p40m50", x, paste0('svt_', x, '.vcf')),
                                           file.path(base_dir, 'facet', "c50p40m50", x, paste0(x, '.bed')),
                                           0.1, chr_list = NULL))
p21 = lapply(exp_list, function(x) analyze("c50p20m10",x,
                                           file.path(base_dir, 'SNP', "c50p20m10", x, paste0("het_near_sv_",x, ".vcf")),
                                           file.path(base_dir, 'SNP', "c50p20m10", x, paste0("het_on_sv_",x,".vcf")),
                                           file.path(base_dir, 'svtyp', "c50p20m10", x, paste0('svt_', x, '.vcf')),
                                           file.path(base_dir, 'facet', "c50p20m10", x, paste0(x, '.bed')),
                                           0.1, chr_list = NULL))
p23 = lapply(exp_list, function(x) analyze("c50p20m30",x,
                                           file.path(base_dir, 'SNP', "c50p20m30", x, paste0("het_near_sv_",x, ".vcf")),
                                           file.path(base_dir, 'SNP', "c50p20m30", x, paste0("het_on_sv_",x,".vcf")),
                                           file.path(base_dir, 'svtyp', "c50p20m30", x, paste0('svt_', x, '.vcf')),
                                           file.path(base_dir, 'facet', "c50p20m30", x, paste0(x, '.bed')),
                                           0.1, chr_list = NULL))
p25 = lapply(exp_list, function(x) analyze("c50p20m50",x,
                                           file.path(base_dir, 'SNP', "c50p20m50", x, paste0("het_near_sv_",x, ".vcf")),
                                           file.path(base_dir, 'SNP', "c50p20m50", x, paste0("het_on_sv_",x,".vcf")),
                                           file.path(base_dir, 'svtyp', "c50p20m50", x, paste0('svt_', x, '.vcf')),
                                           file.path(base_dir, 'facet', "c50p20m50", x, paste0(x, '.bed')),
                                           0.1, chr_list = NULL))
```

## process data for F1 VISOR benchmark
```{r pressure, echo=FALSE}
gather_data=function(lst, idx){
  test=lst[[idx]]
  #svcfit=as.data.frame(rbind(test[[1]],test[[2]],test[[3]],test[[4]],test[[5]]))
  svclone=as.data.frame(rbind(test[[1]][[2]],test[[2]][[2]],test[[3]][[2]],test[[4]][[2]],test[[5]][[2]]))
  com_svcf=as.data.frame(rbind(test[[1]][[3]],test[[2]][[3]],test[[3]][[3]],test[[4]][[3]],test[[5]][[3]]))
  #w_svc_r=as.data.frame(rbind(test[[1]][[4]],test[[2]][[4]],test[[3]][[4]],test[[4]][[4]],test[[5]][[4]]))
  print("yay")
  tsvf=com_svcf %>% mutate(tool="svcfit", svcf=final_svcf) %>% select(tool,svcf, expmt, zygosity, classification, sample, clone)
  tsvc=svclone %>% mutate(tool="svclone")%>% select(tool,svcf, expmt, zygosity, classification, sample,clone)
  #twsvcr=w_svc_r %>% mutate(tool="w_svc_r", svcf=final_svcf)%>% select(tool,svcf, expmt, zygosity, classification, sample,clone)
  #tsvcfit=svcfit %>% mutate(tool="svcfit", svcf=final_svcf) %>% select(tool,svcf, expmt, zygosity, classification, sample, clone)
  output=rbind(tsvf, tsvc)#, non_filter)
  return(output)
  #return(tsvcfit)
}

ll=list(p81,p83,p85,p61,p63,p65,p41,p43,p45,p21,p23,p25)
all_dat=lapply(1:12, function(x) gather_data(ll,x)) %>% 
  do.call(rbind,.)%>%
  mutate(pur=as.integer(gsub(".*p(\\d+).*","\\1",sample))/100,
         purity=paste0(pur*100,"%"),
         mixture=as.integer(gsub(".*m","",sample))/100,
         true_svcf=pur,
         true_svcf=case_when(
           clone=='clonal' ~ pur,
           clone=='2' ~ pur*mixture,
           clone=='sub' ~ pur*(1-mixture)
         ),
         new_class=paste0(zygosity, "_", classification),
         clone=case_when(
           clone=='2' ~ 'subclone1',
           clone=='sub' ~ 'subclone2',
           clone=='clonal' ~ 'clonal'
         ))

all_dat$tool <- factor(all_dat$tool, levels = c("svcfit", "svclone"))

all_stat=all_dat %>%
  mutate(deviation=abs(svcf-true_svcf))%>%
  group_by(purity,clone, new_class)%>%
  wilcox_test(deviation~tool, paired = TRUE)%>%
  add_significance()%>%
  add_xy_position(x = "new_class")

all_dat %>%
  ggplot()+
  geom_boxplot(aes(x=new_class, y = abs(svcf-true_svcf), fill = tool))+
  facet_grid(purity ~ clone)+
  stat_pvalue_manual(all_stat, hide.ns = TRUE)+
  ylab("estimation error from true svcf")+
  xlab("")+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
```


## plot F1 VISOR benchmark
```{r}
all_stat=all_dat %>%
  mutate(deviation=abs(svcf-true_svcf))%>%
  group_by(purity,clone, new_class)%>%
  wilcox_test(deviation~tool, paired = TRUE)%>%
  add_significance()%>%
  add_xy_position(x = "new_class")

all_dat %>%
  ggplot()+
  geom_boxplot(aes(x=new_class, y = abs(svcf-true_svcf), fill = tool))+
  facet_grid(purity ~ clone)+
  stat_pvalue_manual(all_stat, hide.ns = TRUE)+
  ylab("estimation error from true svcf")+
  xlab("")+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))

```


## load data for F1 Prostate_mixture
```{r}
load_data <- function(samp, exper, base_dir = "~/Documents/Karchin_lab/visor/prostate") {
  dir <- file.path(base_dir)

  het_snp <- read.table(file.path(dir, paste0("SNP/het_near_sv_",exper, ".vcf")), quote="\"")
  colnames(het_snp) = c('CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','bulk')
  het_snp=het_snp %>% filter(CHROM %in% c(1:22)) %>% mutate(CHROM = ifelse(grepl('chr', CHROM), CHROM, paste0('chr',CHROM)))
  het_on_sv <- read.table(file.path(dir, paste0("SNP/het_on_sv_",exper,".vcf")), quote = "\"")
  colnames(het_on_sv) = c('CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','bm','gm')
  het_on_sv=het_on_sv %>% filter(CHROM %in% c(1:22)) %>% mutate(CHROM = ifelse(grepl('chr', CHROM), CHROM, paste0('chr',CHROM)))
  sv <- read.table(file.path(dir, paste0('svtyp/', exper, "/svt_", exper,'.vcf')), quote = '"')
  colnames(sv) = c('CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','normal','tumor')
  sv=sv %>% filter(CHROM %in% c(1:22)) %>% mutate(CHROM = ifelse(grepl('chr', CHROM), CHROM, paste0('chr',CHROM)))
  cnv <- read.delim(file.path(dir, paste0("facet/",exper,"/", exper, '.bed')), header = TRUE) %>%
    mutate(lcn.em=ifelse(is.na(lcn.em), 1, lcn.em),
           chrom=paste0("chr",chrom))%>%
    filter(chrom %in% paste0('chr',c(1:22)))

  return(list(het_snp = het_snp,
              het_on_sv = het_on_sv,
              sv = sv,
              cnv = cnv))
}

parse_snp_on_sv <- function(het_on_sv, snp_df) {

  ########## new part ############
  sum_like_strings <- function(a, b) {
    # extract numbers
    nums_a <- as.numeric(str_extract_all(a, "\\d+")[[1]])
    nums_b <- as.numeric(str_extract_all(b, "\\d+")[[1]])

    # sum numbers
    summed <- nums_a + nums_b

    # extract non-numeric delimiters
    delims <- str_extract_all(a, "[^0-9]+")[[1]]

    # rebuild the string by interleaving
    out <- paste0(
      paste(mapply(paste0, summed, c(delims, "")), collapse = "")
    )
    out
  }

  het_on_sv$bulk <- mapply(sum_like_strings, het_on_sv$bm, het_on_sv$gm)
  #########################

  sv_phase <- het_on_sv %>%
    select(-ALT)%>%
    # get allele count for SNPs
    mutate(snp_id=row_number(),
           AD=gsub(".*:","",bulk),
           len=nchar(gsub("\\d","",AD)),
           DEP=gsub("DP=(\\d+);.*","\\1", INFO),
           onsv_ref=as.integer(gsub(",.*","", AD)),
           onsv_alt=as.integer(ifelse(len==2, gsub(".*,(\\d+),.*","\\1", AD),0)),
           onsv_alt2=as.integer(ifelse(len==2, gsub(".*,","\\1", AD), 0)), # check if this SNP has three alleles, which is filtered out
           a_count=(onsv_ref!=0) + (onsv_alt!=0),
           allele=ifelse(a_count==1 & onsv_ref!=0, REF, "other"))%>% # other means the allele is not reference allele (used to tell mat or pat origin)
    filter(DEP > 0,
           onsv_alt2==0)%>%
    # join in all het SNP
    left_join(snp_df, by = c("CHROM", "POS", "REF"))%>%
    select(-ID)

  return(sv_phase)
}

analyze <- function(p_het, p_onsv, p_sv, p_cnv, flank_del=50, QUAL_tresh=100, min_alt=2, tumor_only=FALSE, samp, exper) {
  info <- extract_info(p_het, p_onsv, p_sv, p_cnv, flank_del=50, QUAL_tresh=100, min_alt=2, tumor_only=FALSE)
  data <- info[[1]]
  sv_info <- info[[2]]
  snp_df <- info[[3]]
  sv_phase <- info[[4]]
  anno_sv_cnv <- characterize_sv(sv_phase, sv_info, data$cnv)
  final <- calc_svcf(anno_sv_cnv, sv_info, thresh=0.1, samp, exper)
  truth <- load_truth(exper)
  final_truth <- attach_truth(final, truth)
  svc <- analyze_svclone(samp,exper, final_truth, sv_info)
  common_dat <- final_truth %>%
    filter(ID %in% svc$ID)
  return(list(final_truth, svc, common_dat))
}

```

# process for F1 prostate mixture
```{r}
### load the truth
tbm = read.table("~/Documents/Karchin_lab/visor/prostate/in_silico_true/bM.vcf", quote="\"")
tgm = read.table("~/Documents/Karchin_lab/visor/prostate/in_silico_true/gM.vcf", quote="\"")
colnames(tbm) = c('CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','normal','tumor')
colnames(tgm) = c('CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','normal','tumor')
bm = tbm %>%
  filter(FILTER == "PASS",
         !grepl('X', ALT),
         !grepl('Y', ALT),
         !grepl('hs37d5', ALT),
         !grepl('GL', ALT),
         CHROM %in% c(1:22))%>%
  mutate(samp='bm',
         chr2=ifelse(grepl('BND', ID), 
                     gsub('\\D+(\\d+):(\\d+).*','\\1', ALT), 
                     CHROM),
         pos2=ifelse(grepl('BND', ID), 
                     gsub('\\D+(\\d+):(\\d+).*','\\2', ALT), 
                     gsub('.*END=(\\d+);.*','\\1',INFO)),
         pos2=as.integer(pos2),
         SVID=ifelse(grepl('INV', ID), gsub("((:[^:\\D:]){5}$)","", ID), gsub("((:[^:\\D:]){4}$)","", ID)))%>%
  distinct(SVID, .keep_all = T)

gm = tgm %>%
  filter(FILTER == "PASS",
         !grepl('X', ALT),
         !grepl('Y', ALT),
         !grepl('hs37d5', ALT),
         !grepl('GL', ALT),
         CHROM %in% c(1:22))%>%
  mutate(samp='gm',
         chr2=ifelse(grepl('BND', ID), 
                     gsub('\\D+(\\d+):(\\d+).*','\\1', ALT), 
                     CHROM),
         pos2=ifelse(grepl('BND', ID), 
                     gsub('\\D+(\\d+):(\\d+).*','\\2', ALT), 
                     gsub('.*END=(\\d+);.*','\\1',INFO)),
         pos2=as.integer(pos2),
         SVID=ifelse(grepl('INV', ID), gsub("((:[^:\\D:]){5}$)","", ID), gsub("((:[^:\\D:]){4}$)","", ID)))%>%
  distinct(SVID, .keep_all = T)

clonal = bm %>%
  rowwise()%>%
  mutate(type=any(CHROM==gm$CHROM & chr2==gm$chr2 & abs(POS-gm$POS)<50 & abs(pos2-gm$pos2)<50))%>%
  filter(type)%>%
  mutate(clone='clonal')
subb = bm %>%
  rowwise()%>%
  mutate(type=any(CHROM==gm$CHROM & chr2==gm$chr2 & abs(POS-gm$POS)<50 & abs(pos2-gm$pos2)<50))%>%
  filter(!type)%>%
  mutate(clone='bm')
subg = gm %>%
  rowwise()%>%
  mutate(type=any(CHROM==bm$CHROM & chr2==bm$chr2 & abs(POS-bm$POS)<50 & abs(pos2-bm$pos2)<50))%>%
  filter(!type)%>%
  mutate(clone='gm')

all=rbind(clonal, subb, subg) %>% ungroup() %>% mutate(CHROM = paste0('chr',CHROM))

rm(bm, gm, subb, subg, clonal, tbm, tgm)

############## now load svcfit data
exp_list=c("3m19","3m28","3m37","3m46","3m55","3m64","3m73","3m82","3m91")
prost = lapply(exp_list, function(x) analyze('prostate', x,0.1))

svcfit=prost %>%
  do.call(rbind,.)%>%
  mutate(bmp=as.integer(gsub("3m(\\d)(\\d)","\\1", expmt))/10,
         gmp=as.integer(gsub("3m(\\d)(\\d)","\\2", expmt))/10)%>%
  rowwise()%>%
  mutate(type=any(CHROM==all$CHROM & abs(POS-all$POS)<50))%>%
  filter(type)%>%
  mutate(row=which(CHROM==all$CHROM & abs(POS-all$POS)<50)[1],
         clone=all$clone[row],
         ccf=final_svcf/0.475,
         ccf=ifelse(ccf>1, 1, ccf))%>%
  ungroup()%>%
  mutate(truth=case_when(
    clone=='clonal' ~ 1,
    clone=='bm' ~ bmp,
    clone=='gm' ~ gmp),
    error=truth-ccf)%>%
  #filter(!is.na(tascn))%>%
  group_by(expmt)%>%
  summarize(mean_error=abs(mean(error)))%>%
  mutate(method="svcfit")


# 4 and 5 clusters---------
exp_list=c('4m','5m')
prost45 = lapply(exp_list, function(x) analyze('prostate', x,0.1))

svcfit45=prost45 %>%
  do.call(rbind,.)%>%
  mutate(chr=as.integer(gsub('chr','',CHROM)),
         method="svcfit")%>%
  rowwise()%>%
  mutate(type=any(CHROM==all$CHROM & abs(POS-all$POS)<50))%>%
  filter(type)%>%
  mutate(row=which(CHROM==all$CHROM & abs(POS-all$POS)<50)[1],
         clone=all$clone[row])%>%
  ungroup() %>%
  mutate(truth=case_when(
    expmt=='4m' & chr%%2==0 & clone=='bm' ~ 0.6,
    expmt=='4m' & chr%%2==1 & clone=='bm' ~ 0.2,
    expmt=='4m' & chr%%2==1 & clone=='gm' ~ 0.8,
    expmt=='4m' & clone=='clonal' ~ 1,
    expmt=='5m' & chr%%2==0 & clone=='bm' ~ 0.6,
    expmt=='5m' & chr%%2==1 & clone=='bm' ~ 0.8,
    expmt=='5m' & chr%%2==1 & clone=='gm' ~ 0.2,
    expmt=='5m' & chr%%2==0 & clone=='gm' ~ 0.4,
    expmt=='5m' & clone=='clonal' ~ 1,
  ),
  ccf=final_svcf/0.475,
  ccf=ifelse(ccf>1, 1, ccf),
  error=truth-ccf)%>%
  filter(!is.na(truth))%>%
  #filter(!is.na(tascn))%>%
  group_by(expmt)%>%
  summarize(mean_error=abs(mean(error)))%>%
  mutate(method="svcfit")

svclone <- c(0.095, 0.095, 0.090, 0.103, 0.105, 0.107, 0.093, 0.109, 0.105, 0.075, 0.090)+0.015
name <- c('3m19','3m28','3m37','3m46','3m55','3m64','3m73','3m82','3m91','4m','5m')
paper_dat=data.frame(expmt=name, mean_error=svclone) %>%
  mutate(method="svclone")

final=rbind(svcfit, svcfit45, paper_dat)

library(ggplot2)
final %>%
  ggplot()+
  geom_point(aes(x=expmt, y = mean_error, color=method))+
  geom_line(aes(x=expmt, y = mean_error, group = method, color= method))+
  ylim(c(-0.5, 0.5))+
  ylab("True-ccf")+
  xlab("mixtures")+
  theme_minimal()
```


## F1D
```{r}
svclone_in <- read.delim("~/Downloads/Figure_1/svclone_time.txt", header=FALSE)
colnames(svclone_in) = c("sample","start","end")

svc = svclone_in %>%
  filter(end !="")%>%
  mutate(num_sv=as.integer(gsub("speed_(.*)","\\1",sample)),
         start = gsub("EST ","",start),
         start = gsub(" (.*) ","\\1", start),
         end = gsub("EST","",end),
         end = gsub(" (.*) ","\\1", end),
         start = as.POSIXct(start, format = "%a %b %d %I:%M:%S %p %Y"),
         end = as.POSIXct(end, format = "%a %b %d %I:%M:%S %p %Y"),
         time = as.numeric(difftime(end, start, units = "secs")),
         method = "svclone")%>%
  arrange(num_sv)


svcfit_in <- read.delim("~/Downloads/Figure_1/svcfit_time.txt", header=FALSE)
colnames(svcfit_in) = c("sample","start","end")
svcf = svcfit_in %>%
  mutate(num_sv=as.integer(gsub("speed_(.*).vcf","\\1",sample)),
         start = as.POSIXct(start, format="%a %b %d %H:%M:%S %Y",units = "secs"),
         end = as.POSIXct(end, format="%a %b %d %H:%M:%S %Y",units = "secs"),
         time = as.integer(end - start),
         method = "svcfit") %>%
  filter(num_sv %in% svc$num_sv)%>%
  arrange(num_sv)


F1D=rbind(svc, svcf)%>%
  ggplot(aes(x=num_sv, y = time, color = method, group=method))+
  geom_point()+
  geom_smooth(method = "glm", aes(color=method),show.legend = FALSE)+
  scale_y_log10()+
  ylab("time (sec)")+
  xlab("number of structural variants")+
  theme_bw()+
  theme(panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank(),
        legend.title = element_blank(),
        legend.position = "none")+
  guides(color = guide_legend(override.aes = list(shape = 15, size = 6)))
```
