---
title: "Sup Fig"
author: "Yunzhou Liu"
date: "2025-01-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(SVCFit)
library(dplyr)
library(rstatix)
library(ggplot2)
library(ggsci)
library(stringr)
library(ggh4x)
mendeley_dir = '/Users/lyz928/Documents/mendeley_data'
```

## sequencing depth
```{r}
## all data can be found from mendeley.com, following the instructions from Github
## all data path used here should be in your/pathway/Figure_S

hack <- read.delim(paste0(mendeley_dir, "/VISOR_benchmark/depth/HACk.random.bed"), header=FALSE)
colnames(hack) <- c("chr","left","right","sv","info","flank")
base <- hack %>%
  mutate(id=paste0(chr,"_",left,"_",sv))%>%
  select(chr, left, right, id)%>%
  gather(key="type", value="loc", 2:3)

## expand the locations 2bp to the left and right
## used later to join 
tmp <- lapply(-2:2, function(x) base %>% mutate(loc=loc+x)) %>%
  do.call(rbind,.)

vfp <- list.files(paste0(mendeley_dir, "/VISOR_benchmark/depth"), full.names = T)[1:1001]

read_dep <- function(path){
  test <- read.delim(path, header=FALSE)
  colnames(test) <- c("chr","loc","dep")
  out <- test %>%
    mutate(sample = gsub(".*\\/(.*).txt","\\1",path))
  return(out)
}

vf <- lapply(vfp, function(x) read_dep(x))%>%
  do.call(rbind,.)%>%
  inner_join(tmp)%>%
  group_by(id,type,sample)%>%
  summarise(fin_dep = max(dep))

#pur <- c(90,80,70,60,50,40,30,20,10,100)

dat_start <- vf %>%
  filter(type == "left")%>%
  rowwise()%>%
  mutate(purity = as.numeric(gsub(".*p(\\d+).*","\\1",sample))/100)%>%
  group_by(purity)%>%
  mutate(ave = mean(fin_dep))%>%
  ungroup()

dat_end <- vf %>%
  filter(type == "right")%>%
  rowwise()%>%
  mutate(purity = as.numeric(gsub(".*p(\\d+).*","\\1",sample))/100)%>%
  group_by(purity)%>%
  mutate(ave = mean(fin_dep))%>%
  ungroup()

dat <- vf %>%
  rowwise()%>%
  mutate(purity = as.numeric(gsub(".*p(\\d+).*","\\1",sample))/100)%>%
  group_by(purity)%>%
  mutate(ave = mean(fin_dep))%>%
  ungroup()%>%
  pivot_wider(names_from = type, values_from = fin_dep)

corl <- round(cor(x=dat_start$purity,y=dat_start$fin_dep, method = "pearson"), digits = 2)
corr <- round(cor(x=dat_end$purity,y=dat_end$fin_dep, method = "pearson"), digit=2)
```


```{r}
p1=dat_start%>%
  ggplot(aes(x=purity, y=fin_dep))+
  geom_point(size = 2, alpha = .3)+
  geom_smooth(method="glm", show.legend = T)+
  annotate("text", x=0.3, y=85, label= paste0("pearson corr: ", corl), size = 15/.pt)+
  xlab("SV purity level (%)")+
  ylab("read depth")+
  theme_classic()

p2=dat_end%>%
  ggplot(aes(x=purity, y=fin_dep))+
  geom_point(size = 2, alpha = .3)+
  geom_smooth(method="glm", show.legend = T)+
  annotate("text", x=0.3, y=85, label= paste0("pearson corr: ", corr), size = 15/.pt)+
  xlab("SV purity level (%)")+
  ylab("read depth")+
  theme_classic()

```

## Fig. S4
```{r}
### all_dat is produced in paper/Figure_1.Rmd
# --- STEP 1: PREPARE DATA & CLEAN NAMES ---
clean_dat <- all_dat %>%
  filter(expmt == 'exp5') %>%
  mutate(
    # Create the "Deviation" column for the test
    deviation = abs(svcf - true_svcf),
    
    # Clean up Classification (X-axis)
    classification = case_when(
      classification == 'INV' ~ 'Inversion',
      classification == 'DUP' ~ 'Tandem Duplication',
      classification == 'DEL' ~ 'Deletion',
      classification == 'BND' ~ 'Translocation',
      TRUE ~ classification
    ),
    # Clean up Zygosity (Facet Header)
    zygosity = case_when(
      zygosity == 'het' ~ 'Heterozygous',
      zygosity == 'hom' ~ 'Homozygous',
      TRUE ~ zygosity
    )
  )

# --- STEP 2: CALCULATE STATS ---
tmp_stat <- clean_dat %>%
  # Group by the NEW structure (separated zygosity)
  group_by(purity, clone, zygosity, classification) %>%
  
  wilcox_test(deviation ~ tool, paired = TRUE) %>%
  add_significance() %>%
  
  # CRITICAL CHANGE: Tell it the X-axis is now 'classification'
  add_xy_position(x = "classification")


# --- STEP 

x = clean_dat %>%
  ggplot(aes(x = classification, y = deviation)) +
  
  geom_boxplot(aes(fill = tool), outlier.size = 0.5, outlier.alpha = 0.5, lwd = 0.4) +
  scale_fill_npg() + 
  
  facet_nested(purity ~ clone + zygosity, scales = "free_x", space = "free_x") +
  
  stat_pvalue_manual(tmp_stat, hide.ns = TRUE, label = "p.signif", tip.length = 0.01) +
  
  #scale_x_discrete(labels = function(x) str_wrap(x, width = 10)) +
  
  ylab("Absolute SVCF estimation error") +
  xlab(NULL) +
  
  theme_classic(base_size = 14) + 
  theme(
    text = element_text(family = "sans"),
    
    # --- CRITICAL FIX START ---
    # 1. Turn OFF the default axis lines (Fixes the "double thickness" on left/bottom)
    axis.line = element_blank(),
    
    # 2. Draw a uniform box around the plot area
    panel.border = element_rect(color = "black", fill = NA, size = 0.4),
    
    # 3. Draw a uniform box around the headers (Clone/Zygosity)
    strip.background = element_rect(fill = "white", color = "black", size = 0.3),
    # --- CRITICAL FIX END ---
    
    # Spacing and Grid
    panel.spacing = unit(0.3, "lines"), 
    panel.grid.major.y = element_line(color = "grey92", size = 0.3),
    panel.grid.major.x = element_blank(),
    
    # Axis Text
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, color = "black", size = 10),
    strip.text = element_text(face = "bold", size = 10),
    
    legend.position = "bottom",
    legend.title = element_blank()
  )
```



```{r}
plot_circos <- function(samp, sv_path, cnv_path, svcfit_path, cyto_path, 
                        cluster_path, pair_num, shared=FALSE, pre_sample=TRUE) {
  
  # 1. Setup Cytoband
  cyto <- read.delim(cyto_path, header=FALSE)
  colnames(cyto) <- c("chrom", "start", "end", "name", "gieStain")
  chr_lst <- paste0("chr", c(1:22))
  cyto <- cyto %>% filter(chrom %in% chr_lst)
  
  # 2. Prepare SV Info
  # Read sub-selection BED if exists
  svcfit_dat <- read.delim(svcfit_path)
  
  if(shared){
    clust_dat <- read.csv(cluster_path) %>%
    mutate(type=ifelse(pre_BAT*on_BAT!=0, 'shared', 'unique'))%>%
    filter(pair==pair_num, type=='shared')
  }else if (!shared & pre_sample){
    clust_dat <- read.csv(cluster_path) %>%
    mutate(type=ifelse(pre_BAT*on_BAT!=0, 'shared', 'unique'))%>%
    filter(pair==pair_num, type=='unique', on_BAT==0)
  }else{
    clust_dat <- read.csv(cluster_path) %>%
    mutate(type=ifelse(pre_BAT*on_BAT!=0, 'shared', 'unique'))%>%
    filter(pair==pair_num, type=='unique', pre_BAT==0)
  }
  
  sv <- read.table(sv_path, quote="\"")
  colnames(sv) <- c('CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','tumor')
  
  sv_info <- sv %>%
    mutate(
      chr2 = case_when(
        grepl('BND', ID) ~ gsub('.*(chr.*):.*','\\1',ALT),
        !grepl('BND', ID) ~ CHROM
      ),
      END = case_when(
        grepl('BND', ID) ~ gsub('.*chr\\w+:(\\d+).*','\\1',ALT),
        !grepl('BND', ID) ~ gsub('.*;END=(\\d+);.*','\\1', INFO)
      ),
      svtype = gsub('.*;SVTYPE=(\\w+);.*','\\1', INFO),
      POS = as.integer(POS),
      END = as.integer(END)
    ) %>%
    filter(CHROM %in% chr_lst,
           chr2 %in% chr_lst,
           FILTER == 'PASS',
           ID %in% svcfit_dat$ID)  %>% # Filter by the BED file IDs
    left_join(clust_dat)%>%
    filter(!is.na(cluster_num))
  
  # 3. Prepare CNV Info
  cnv <- read.delim(cnv_path)
  cnv_info <- cnv %>%
    mutate(tcn.em = ifelse(tcn.em == 0, 0.001, tcn.em),
           log2 = round(log2(tcn.em/2), digits = 3),
           start = as.integer(start),
           end = as.integer(end),
           chrom = paste0('chr', chrom)) %>%
    select(chrom, start, end, log2) %>%
    filter(chrom %in% chr_lst)
  
  # 4. Plotting
  par(cex = 1.4)
  circos.clear()
  circos.par(gap.after = c(rep(1, length(chr_lst)-1), 6), start.degree = 90, track.margin = c(0.01, 0.01))
  
  # Initialize
  circos.initializeWithIdeogram(cyto, plotType = c("ideogram", "labels"))
  
  # Track: CNV segments
  col_fun <- colorRamp2(c(-2, 0, 2), c("#4575b4", "grey90", "#d73027"))
  circos.genomicTrack(
    cnv_info, ylim = c(-2, 2), track.height = 0.12, bg.border = NA,
    panel.fun = function(region, value, ...) {
      circos.genomicRect(region, value, col = col_fun(value$log2), border = NA, ...)
      circos.lines(CELL_META$xlim, c(0, 0), col = "#555555", lwd = 0.5)
    }
  )
  
  # Links: SVs
  sv_cols <- c(DEL = "#1f78b4", INV = "#33a02c", DUP = "#ff7f00", BND = "#6a3d9a")
  sv_cols_mapped <- sv_cols[sv_info$svtype]
  sv_cols_mapped[is.na(sv_cols_mapped)] <- "#e31a1c"
  
  links1 <- data.frame(chr = sv_info$CHROM, start = sv_info$POS, end = sv_info$POS, stringsAsFactors = FALSE)
  links2 <- data.frame(chr = sv_info$chr2, start = sv_info$END, end = sv_info$END, stringsAsFactors = FALSE)
  
  circos.genomicLink(
    links1, links2,
    col = adjustcolor(sv_cols_mapped, alpha.f = 0.7),
    border = NA
  )
  
  legend("topright", legend = c('deletion','inversion', 'tandem duplication', 'translocation'), col = sv_cols, pch = 15, bty = "n")
}

```

# F3d: circus plot 
```{r}
samp = "81875"
pair=4
sv_path=paste0(mendeley_dir, '/COMBAT/sv_call/survivor/svcfit_', samp, '.vcf')
cnv_files=list.files(paste0(mendeley_dir, '/COMBAT/facets'), full.names = T)
cnv_path=cnv_files[grep(samp, cnv_files)]
svcfit_path=paste0(mendeley_dir, '/COMBAT/SVCFit_output/', samp, '.bed')
cyto_path=paste0(mendeley_dir, '/COMBAT/cytoBand.txt')
cluster_path=paste0(mendeley_dir, '/COMBAT/sv2cluster.csv')

#SHARED
plot_circos(samp, sv_path, cnv_path, svcfit_path, cyto_path, cluster_path, pair_num=pair, shared=T, pre_sample = F)
#PRE-BAT
plot_circos(samp, sv_path, cnv_path, svcfit_path, cyto_path, cluster_path, pair_num=pair, shared=F, pre_sample = T)


samp = "82780_recut"
pair=4
sv_path=paste0(mendeley_dir, '/COMBAT/sv_call/survivor/svcfit_', samp, '.vcf')
cnv_files=list.files(paste0(mendeley_dir, '/COMBAT/facets'), full.names = T)
cnv_path=cnv_files[grep(samp, cnv_files)]
svcfit_path=paste0(mendeley_dir, '/COMBAT/SVCFit_output/', samp, '.bed')
cyto_path=paste0(mendeley_dir, '/COMBAT/cytoBand.txt')
cluster_path=paste0(mendeley_dir, '/COMBAT/sv2cluster.csv')
#ON-BAT
plot_circos(samp, sv_path, cnv_path, svcfit_path, cyto_path, cluster_path, pair_num=pair, shared=F, pre_sample = F)


samp = "85293"
pair=10
sv_path=paste0(mendeley_dir, '/COMBAT/sv_call/survivor/svcfit_', samp, '.vcf')
cnv_files=list.files(paste0(mendeley_dir, '/COMBAT/facets'), full.names = T)
cnv_path=cnv_files[grep(samp, cnv_files)]
svcfit_path=paste0(mendeley_dir, '/COMBAT/SVCFit_output/', samp, '.bed')
cyto_path=paste0(mendeley_dir, '/COMBAT/cytoBand.txt')
cluster_path=paste0(mendeley_dir, '/COMBAT/sv2cluster.csv')

#SHARED
plot_circos(samp, sv_path, cnv_path, svcfit_path, cyto_path, cluster_path, pair_num=pair, shared=T, pre_sample = F)
#PRE-BAT
pdf("/Users/lyz928/Karchin\ Lab\ Dropbox/YunZhou\ Liu/SVCFit-2024/NPJ_Precision_Oncology_submission/figures/subject31_pre.pdf", width = 10, height = 10)
plot_circos(samp, sv_path, cnv_path, svcfit_path, cyto_path, cluster_path, pair_num=pair, shared=F, pre_sample = T)


samp = "86921"
pair=10
sv_path=paste0(mendeley_dir, '/COMBAT/sv_call/survivor/svcfit_', samp, '.vcf')
cnv_files=list.files(paste0(mendeley_dir, '/COMBAT/facets'), full.names = T)
cnv_path=cnv_files[grep(samp, cnv_files)]
svcfit_path=paste0(mendeley_dir, '/COMBAT/SVCFit_output/', samp, '.bed')
cyto_path=paste0(mendeley_dir, '/COMBAT/cytoBand.txt')
cluster_path=paste0(mendeley_dir, '/COMBAT/sv2cluster.csv')
#ON-BAT
pdf("/Users/lyz928/Karchin\ Lab\ Dropbox/YunZhou\ Liu/SVCFit-2024/NPJ_Precision_Oncology_submission/figures/subject31_on.pdf", width = 10, height = 10)
plot_circos(samp, sv_path, cnv_path, svcfit_path, cyto_path, cluster_path, pair_num=pair, shared=F, pre_sample = F)
```
